[PATCH] TOMOYO: Fix race on updating profile's comment line.

commit 2a086e5d3a23570735f75b784d29b93068070833 upstream.

tomoyo_save_name() in tomoyo_write_profile() may return NULL.
Therefore,

	profile->comment ? profile->comment->name : ""

in tomoyo_read_profile() may race.
Keep the old value rather than replace with empty string when out of memory
error has occurred.

Signed-off-by: Xiaochen Wang <wangxiaochen0@gmail.com>
Signed-off-by: Tetsuo Handa <penguin-kernel@I-love.SAKURA.ne.jp>
---
 security/tomoyo/common.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- linux-2.6.32.40.orig/security/tomoyo/common.c
+++ linux-2.6.32.40/security/tomoyo/common.c
@@ -924,7 +924,11 @@ static int tomoyo_write_profile(struct t
 		return -EINVAL;
 	*cp = '\0';
 	if (!strcmp(data, "COMMENT")) {
-		profile->comment = tomoyo_save_name(cp + 1);
+		const struct tomoyo_path_info *new_comment
+			= tomoyo_save_name(cp + 1);
+		if (!new_comment)
+			return -ENOMEM;
+		profile->comment = new_comment;
 		return 0;
 	}
 	for (i = 0; i < TOMOYO_MAX_CONTROL_INDEX; i++) {
